networkx/algorithms/planarity.py

```diff
@@ -858,6 +855,22 @@ class PlanarEmbedding(nx.DiGraph):
 
     """
 
+    def __init__(self, incoming_graph_data=None, **attr):
+        super().__init__(incoming_graph_data=incoming_graph_data, **attr)
+        self.add_edge = self.__forbidden
+        self.add_edges_from = self.__forbidden
+        self.add_weighted_edges_from = self.__forbidden
+
+    def __forbidden(self, *args, **kwargs):
+        """Forbidden operation
+
+        Any edge additions to a PlanarEmbedding should be done using
+        method `add_half_edge`.
+        """
+        raise nx.NetworkXError(
+            "Use `add_half_edge` method to add edges to a PlanarEmbedding."
+        )
+
```

Comment:
I may be missing something here, but IMO it'd be more idiomatic to override these methods directly instead of assigning them to a hidden/forbidden method, e.g.

```python

def add_edge(self, *args, **kwargs):
    raise NotImplemented("Use `add_half_edge` to add edges to a PlanarEmbedding")

# same for the other methods
```

I'd also vote for using `NotImplemented` instead of `NetworkXError`, which I think fits slightly better and will automatically resolve the "tests passing for the wrong reason" issue.

Reply:
Thanks for reviewing, @rossbar 

I rebased again this branch and now I think I stacked on top of PR #7202 correctly.

 > I may be missing something here, but IMO it'd be more idiomatic to override these methods directly instead of assigning them to a hidden/forbidden method, e.g.

The constructor of *DiGraph* (*PlanarEmbedding*'s parent class) uses `self.add_edge*()`. This happens via 
delegation of *incoming_graph_data* to functions in `convert.py`. If I override `add_edge*()` in the class definition, those functions will fail. I avoided that by extending *DiGraph*'s constructor and overriding `add_edge*()` after *incoming_graph_data* is processed.

I think that in order to use the idiomatic approach you suggested, the *PlanarEmbedding*'s constructor would have to do something like that if it receives graph data:

```python
#  not exactly what we need, since it returns a new object
LRPlanarity(nx.Graph(incoming_graph_data)).lr_planarity() 
```

Though this would make it very hard to create invalid embeddings for the test instances.


> I'd also vote for using `NotImplemented` instead of `NetworkXError`, which I think fits slightly better and will automatically resolve the "tests passing for the wrong reason" issue.

The "tests passing for the wrong reason" issue was already solved by only using the context `pytest.raises()` for the actual functionality under test.

Let me know if the difficulties with checking this PR out remains.

Reply:
Ah, good point - thanks for the explanation, that makes sense. It sounds like there's opportunity for a deeper refactor here, but I'm inclined not to let that be a blocker for this PR!

> The "tests passing for the wrong reason" issue was already solved by only using the context pytest.raises() for the actual functionality under test.

True, though I still feel like `NotImplemented` is a better-fitting exception, as this has to do with methods on the Python object we're forbidding, rather than anything dealing with graphs (in the abstract sense), where a NetworkX-flavored exception would be expected. This is very subjective though and also not a blocker!